### ISSUE TITLE
Create a Tcp Server/Clientfor the MultiNode Testkit

### ISSUE THREAD

**Issue by rogeralsing** [akkadotnet/akka.net#1475]
As discussed on the last Core dev meeting, we should replace the current StdOut logging for MultiNode Testkit to instead log using a custom logger that sends all events over Tcp using Akka.IO.

So we need a Tcp server that aggregates all the events, and for each node we need a custom logger that can connect and talk  to the server.

@Aaronontheweb and @Horusiath has more details about this

**Comment by maxim-s** [comment]
@Aaronontheweb  and @Horusiath could you describe a bit more?

**Comment by Aaronontheweb** [comment]
sure thing @maxim-s - this is one of the most important things we need done in order to ship Akka.NET 1.1.

TL;DR; - this is how the multi-node testkit communicates between processes right now:

```csharp
foreach (var nodeTest in test.Value)
                        {
                            //Loop through each test, work out number of nodes to run on and kick off process
                            var process = new Process();
                            processes.Add(process);
                            process.StartInfo.UseShellExecute = false;
                            process.StartInfo.RedirectStandardOutput = true;
                            process.StartInfo.FileName = "Akka.NodeTestRunner.exe";
                            process.StartInfo.Arguments = String.Format(@"-Dmultinode.test-assembly=""{0}"" -Dmultinode.test-class=""{1}"" -Dmultinode.test-method=""{2}"" -Dmultinode.max-nodes={3} -Dmultinode.server-host=""{4}"" -Dmultinode.host=""{5}"" -Dmultinode.index={6}",
                                assemblyName, nodeTest.TypeName, nodeTest.MethodName, test.Value.Count, "localhost", "localhost", nodeTest.Node - 1);
                            var nodeIndex = nodeTest.Node;
                            process.OutputDataReceived +=
                                (sender, line) =>
                                {
                                    //ignore any trailing whitespace
                                    if (string.IsNullOrEmpty(line.Data) || string.IsNullOrWhiteSpace(line.Data)) return;
                                    string message = line.Data;
                                    if (!message.StartsWith("[NODE", true, CultureInfo.InvariantCulture))
                                    {
                                        message = "[NODE" + nodeIndex + "]" + message;
                                    }
                                    PublishToAllSinks(message);
                                };

                            var closureTest = nodeTest;
                            process.Exited += (sender, eventArgs) =>
                            {
                                if (process.ExitCode == 0)
                                {
                                    ReportSpecPassFromExitCode(nodeIndex, closureTest.TestName);
                                }
                            };
                            process.Start();

                            process.BeginOutputReadLine();
```

We're using STD OUT redirection, which means each of the data we receive back from all of the individual child processes arrives in 80 character wide lines. So in order to rebuild these logs into something intelligible on the MultiNodeTestRunner.exe we have to use a set of gnarly regular expressions to glue all of the stacktraces and logs from individual nodes back together into something intelligible.

Due to the inherently racy nature of having 1-6 processes redirect their output at once, we end up not being able to put many of those stacktraces and logs back together properly. Therefore, we discard them - which means we lose data we need to be able to debug failures that can occur.

So the solution then is to strip out STD IO redirection entirely and replace it with a TCP / UDP (probably UDP would be best) system on top of Akka.IO so we can capture the entire C# object that included the log message or stacktrace, rather than have to parse it from strings.

Does that help make things more clear?

**Comment by maxim-s** [comment]
@Aaronontheweb thank you, it's clear

**Comment by Aaronontheweb** [comment]
@maxim-s hey Maxim, do you have any code on this yet?

**Comment by maxim-s** [comment]
@Aaronontheweb Hello, I've finished digging in #1365 , and  I'm going to start with this issue today, sorry for delay

**Comment by Aaronontheweb** [comment]
@maxim-s no worries Maxim - would it be alright if I took this one on instead? This just moved up to the top of my "urgent" list

**Comment by Aaronontheweb** [comment]
Proposed design for this:

http://cloud.smartdraw.com/share.aspx/?pubDocShare=F42B9EE6FD6054DE5DD52CDA1A7E5DE01D0

Essentially we're going to implement a custom Akka.NET `Logger` that writes out to a configured Akka.IO UDP or TCP transport, and that destination address information will be passed into the arguments of each individual `NodeTestRunner.exe` process via their startup parameters.

**Comment by maxim-s** [comment]
@Aaronontheweb yes sure, but I want something instead :)

**Comment by Aaronontheweb** [comment]
You got it - if you're already done looking into #1465 (please add some comments on the thread).

Others that can use some love:
#1431 
#1089 
And any of these issues assigned to @willieferguson, whom I haven't heard from in a while: https://github.com/akkadotnet/akka.net/issues?q=assignee%3Awillieferguson+is%3Aopen
